JSON configuration for `cchost` (#3209)
microsoft/CCF: doc/operations/configuration.rst

> Configuration

Reviewer: Comment: We've leaned hard into using `configuration` to describe sets of active nodes, and the adding/removing node protocols. This was fine while we started a node with a set of "options", but is confusing if we start a node with a "configuration". But it's going to confuse somebody that `configuration.rst` has nothing to do with `reconfiguration.rst` (or currently, `1tx-reconfig.rst` and `2tx-reconfig.rst`). In particular, if we dump a portion of this configuration file inside the KV, and think some of these settings are updateable by governance, what is _that_ called?

To be clear, I think this new use is sensible and absolutely the right term, matching common uses of the configure/configuration elsewhere, and I'm not suggesting a rename.

Author: Agreed. As discussed, there isn't an obvious alternative name here ("membership" is already used for consortium members) - perhaps "servicing a network" for what we currently call node reconfiguration? I'm happy to revisit this later and upgrade our docs once #2612 is done.

=============

The configuration for each CCF node must be contained in a single JSON configuration file specified to the ``cchost`` executable via the ``--config </path/to/configuration/file>`` argument.

.. tip:: JSON configuration samples:

    - Minimal configuration: https://github.com/microsoft/CCF/blob/main/samples/config/minimal_config.json
    - Full configuration: https://github.com/microsoft/CCF/blob/main/samples/config/config.json

Configuration Options
---------------------

``enclave_file``
~~~~~~~~~~~~~~~~

Path to CCF enclave application.

``enclave_type``
~~~~~~~~~~~~~~~~

> Type of enclave application (either ``release``, ``debug`` or ``virtual``). Default value: ``release``.

Reviewer 2: Every time I see several `foo_(a|b|..)`, I wonder if we shouldn't have a section `foo` with `foo.a`, `foo.b`. On the one hand, it seems cleaner and more future proof (can add fields without crowding the top level), but on the other hand, how much nesting is too much nesting? I genuinely do not know.

``network``
~~~~~~~~~~~

The ``network`` section includes configuration for the interfaces a node listens on (both node-to-node and RPC).

- ``node_address``: Address (hostname and port) to listen on for incoming node-to-node connections.

- ``rpc_interfaces``: Addresses (hostname and port) to listen on for incoming client TLS connections.
Each RPC address must contain:

> - The local RPC address (``rpc_address``) the node listens on.

Reviewer 2: I wonder if we should go for `bind_address` and `advertised_address`?

Author: Done. Renamed to `bind_address` and `published_address`.

- ``initial_validity_period_days``: Initial validity period (days) for node certificate. Default value: ``1`` day.

.. _start configuration:

``start``
~~~~~~~~~

.. note:: This only needs to be set when the node started in ``start`` mode.

- ``constitution_files``: List of constitution files. These typically include ``actions.js``, ``validate.js``, ``resolve.js`` and ``apply.js``.

- ``members``: List of initial consortium members files, including identity certificates, public encryption keys and member data files.

.. note:: Common examples:

    - A recovery member with member data: ``{"certificate_file": "member_cert.pem", "encryption_public_key_file": "member_enc_pubk.pem", "data_json_file": "member_data.json"}``
    - A recovery member with no member data: ``{"certificate_file": "member_cert.pem", "encryption_public_key_file": "member_enc_pubk.pem"}``
    - A non-recovery member with member data: ``{"certificate_file": "member_cert.pem", "data_json_file": "member_data.json"}``
    - A non-recovery member with no member data: ``{"certificate_file": "member_cert.pem"}``

- ``service_configuration``: Initial service configuration, including:

>     - ``max_allowed_node_cert_validity_days``: The maximum number of days allowed for node certificate validity period. Default value: ``365`` days. TODO: Rename

Reviewer 2: rename?

    - ``recovery_threshold``. Note that if the recovery threshold is set to ``0``, it is automatically set to the number of recovery members specified in ``members``.
    - ``reconfiguration_type``. The type of reconfiguration for new nodes. Default value: ``OneTransaction``.

Example:

.. code-block:: json

    "start": {
        "constitution_files": ["actions.js", "validate.js", "resolve.js", "apply.js"],
        "members": [
            {"certificate_file": "member0_cert.pem", "data_json_file": null, "encryption_public_key_file": "member0_enc_pubk.pem"},
            {"certificate_file": "member1_cert.pem", "data_json_file": "member1_data.json", "encryption_public_key_file": null},

>             {"certificate_file": "member2_cert.pem", "data_json_file": null, "encryption_public_key_file": "member2_enc_pubk.pem"}

Reviewer 2: We can probably omit the keys for null values altogether?

        ],
        "service_configuration":
        {
            "recovery_threshold": 0,
            "max_allowed_node_cert_validity_days": 365,
            "reconfiguration_type": "OneTransaction"
        }
    }

.. _join configuration:

``join``
~~~~~~~~

.. note:: This only needs to be set when the node is started in ``join`` mode.

- ``target_rpc_address``: Address (hostname and port) of a node of the existing service to join.
- ``join_timer_ms``: Interval (ms) at which the node sends join requests to the existing network. Default value: ``1000`` ms.

Example:

.. code-block:: json

    "join": {
        "join_timer_ms": 1000,
        "target_rpc_address": {"hostname": "127.0.0.1", "port": "8080"}
    }

``ledger``
~~~~~~~~~~

> - ``ledger_dir``: Path to main ledger directory. Default value: ``ledger``.

Reviewer 2: `ledger` feels redundant here, perhaps `append_directory` and `read_only_directories` (and `chunk_size_bytes`?)

Author: `chunk_size` as the unit will be in value string.

- ``read_only_ledger_dirs``: Optional. Paths to read-only ledger directories. Note that only ``.committed`` files will be read from these directories. Default value: ``[]``.
- ``ledger_chunk_bytes``: Minimum size (bytes) of the current ledger file after which a new ledger file (chunk) is created. Default value: ``5000000`` bytes.

``snapshots``
~~~~~~~~~~~~~

- ``snapshots_dir``: Path to snapshot directory. Default value: ``snapshots``.

> - ``snapshot_tx_interval``: Minimum number of transactions between snapshots. Default value: ``10000``.

Reviewer 2: Same as for ledger, these names feel redundant, and `interval_size_tx` would be somewhat more consistent with other options (unit at the end).

``logging``
~~~~~~~~~~~

- ``host_log_level``: Logging level for the `untrusted host`. Default value: ``INFO``.

> .. note:: While it is possible to set the host log level at startup, it is not possible to change the log level of the enclave.

Reviewer 2: ```suggestion
.. note:: While it is possible to set the host log level at startup, it is not possible to change the log level of the enclave without rebuilding it.
```

> - ``log_format_json``: If ``true``, node logs will be formatted as JSON. Default value: ``false``.

Reviewer 2: `logging.format` as `text|json` seems a little more flexible and concise?

``consensus``
~~~~~~~~~~~~~

> - ``type``: Type of consensus protocol. Only ``CFT`` (Crash-Fault Tolerant) is currently supported in production. Default value: ``CFT``.

Reviewer 2: ```suggestion
- ``type``: Type of consensus protocol. Only ``CFT`` (Crash-Fault Tolerant) is currently supported. Default value: ``CFT``.
```

> - ``timeout_ms``: Hearbeat interval (ms) at which primary node sends messages to backup nodes to maintain primary-ship. This should be set to a significantly lower value than ``election_timeout_ms``. Default value: ``100`` ms.

Reviewer 2: Heartbeats aren't quite the whole story, it's also how often the primary retries sending messages that haven't been acked.

> - ``election_timeout_ms``: Timeout value (ms) after which backup node that have not received primary heartbeats will trigger a new election. Default timeout: ``4000`` ms.

Reviewer 2: ```suggestion
- ``election_timeout_ms``: Timeout value (ms) after which backup node that have not received any message from the primary will trigger a new election. Default timeout: ``4000`` ms.
```

Reviewer 2: Also probably worth mentioning that this should be set to be substantially larger than `timeout_ms`.

``intervals``
~~~~~~~~~~~~~

- ``sig_tx_interval``: Number of transactions after which a signature transaction is automatically generated. Default value: ``5000``.
- ``sig_ms_interval``: Maximum duration (milliseconds) after which a signature transaction is automatically triggered. Default value: ``1000`` ms.

.. note::
    Transaction commit latency in a CCF network is primarily a function of signature frequency. A network emitting signatures more frequently will be able to commit transactions faster, but will spend a larger proportion of its execution resources creating and verifying signatures. Setting signature frequency is a trade-off between transaction latency and throughput.

>     The signature interval options specify the intervals at which the generation of signature transactions is `triggered`. However, because of the parallel execution of transactions, the actual intervals between signature transactions may be slightly larger.

Reviewer 2: It's not just parallel execution, it's also queuing.
